op-node/rollup/interop/interop.go

```diff
@@ -0,0 +1,159 @@
+package interop
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/engine"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/event"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+const checkBlockTimeout = time.Second * 10
+
+type InteropBackend interface {
+	CheckBlock(ctx context.Context,
+		chainID types.ChainID, blockHash common.Hash, blockNumber uint64) (types.SafetyLevel, error)
+}
+
+type L2Source interface {
+	L2BlockRefByNumber(context.Context, uint64) (eth.L2BlockRef, error)
+}
+
+// InteropDeriver watches for update events (either real changes to block safety,
+// or updates published upon request), checks if there is some local data to cross-verify,
+// and then checks with the interop-backend, to try to promote to cross-verified safety.
+type InteropDeriver struct {
+	log log.Logger
+	cfg *rollup.Config
+
+	// we cache the chainID,
+	// to not continuously convert from the type in the rollup-config to this type.
+	chainID types.ChainID
+
+	driverCtx context.Context
+
+	// L2 blockhash -> derived from L1 block ref.
+	// Added to when a block is local-safe.
+	// Removed from when it is promoted to cross-safe.
+	derivedFrom map[common.Hash]eth.L1BlockRef
+
+	backend InteropBackend
+	l2      L2Source
+
+	emitter event.Emitter
+
+	mu sync.Mutex
+}
+
+var _ event.Deriver = (*InteropDeriver)(nil)
+var _ event.AttachEmitter = (*InteropDeriver)(nil)
+
+func NewInteropDeriver(log log.Logger, cfg *rollup.Config,
+	driverCtx context.Context, backend InteropBackend, l2 L2Source) *InteropDeriver {
+	return &InteropDeriver{
+		log:         log,
+		cfg:         cfg,
+		chainID:     types.ChainIDFromBig(cfg.L2ChainID),
+		driverCtx:   driverCtx,
+		derivedFrom: make(map[common.Hash]eth.L1BlockRef),
+		backend:     backend,
+		l2:          l2,
+	}
+}
+
+func (d *InteropDeriver) AttachEmitter(em event.Emitter) {
+	d.emitter = em
+}
+
+func (d *InteropDeriver) OnEvent(ev event.Event) bool {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	switch x := ev.(type) {
+	case engine.UnsafeUpdateEvent:
+		d.emitter.Emit(engine.RequestCrossUnsafeEvent{})
+	case engine.CrossUnsafeUpdateEvent:
+		if x.CrossUnsafe.Number >= x.LocalUnsafe.Number {
+			break // nothing left to promote
+		}
+		// pre-interop the engine itself handles promotion to cross-unsafe
+		if !d.cfg.IsInterop(d.cfg.TimestampForBlock(x.CrossUnsafe.Number + 1)) {
+			return false
+		}
+		ctx, cancel := context.WithTimeout(d.driverCtx, checkBlockTimeout)
+		defer cancel()
+		candidate, err := d.l2.L2BlockRefByNumber(ctx, x.CrossUnsafe.Number+1)
```

Comment:
to match the spacing in the prior case
```suggestion
		candidate, err := d.l2.L2BlockRefByNumber(ctx, x.CrossUnsafe.Number + 1)
```

Reply:
The Go formatter doesn't like the `space`, because there are multiple function arguments, and the no-space sum statement groups it clearly as argument.